clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -471,10 +547,31 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
     /**
      * Record any newly completed connections
      */
-    private void handleConnections() {
+    private void handleConnections(long now) {
         for (String node : this.selector.connected()) {
             log.debug("Completed connection to node {}", node);
             this.connectionStates.connected(node);
+            // Though the node is connected, we might not still be able to send requests. For instance,
+            // in case of SSL connection, SSL handshake happens after connection is established.
+            if (!apiVersionsRequestQueue.contains(node))
+                apiVersionsRequestQueue.add(node);
+        }
+        maybeInitiateApiVersionsFetch(now);
+    }
+
+    private void maybeInitiateApiVersionsFetch(long now) {
+        if (this.requiredApiVersions == null)
+            return;
+
+        for (String node : apiVersionsRequestQueue) {
+            if (selector.isChannelReady(node) && inFlightRequests.canSendMore(node)) {
+                log.debug("Initiating API versions fetch from node {}.", node);
+                this.connectionStates.checkingApiVersions(node);
+                ApiVersionsRequest apiVersionRequest = new ApiVersionsRequest();
+                RequestSend send = new RequestSend(node, nextRequestHeader(ApiKeys.API_VERSIONS), apiVersionRequest.toStruct());
+                doSend(new ClientRequest(now, true, send, null, true), now);
```

Comment:
why not use a future + callback handler? The inline check in `handleCompletedReceives` is a bit confusing / buried